README.md
# sessionKeys (Ruby)

[![Gem Version](https://badge.fury.io/rb/session_keys.svg)](https://badge.fury.io/rb/session_keys)
[![Build Status](https://travis-ci.org/grempe/session-keys-rb.svg?branch=master)](https://travis-ci.org/grempe/session-keys-rb)
[![Coverage Status](https://coveralls.io/repos/github/grempe/session-keys-rb/badge.svg?branch=master)](https://coveralls.io/github/grempe/session-keys-rb?branch=master)
[![Code Climate](https://codeclimate.com/github/grempe/session-keys-rb/badges/gpa.svg)](https://codeclimate.com/github/grempe/session-keys-rb)
[![Inline docs](http://inch-ci.org/github/grempe/session-keys-rb.svg?branch=master)](http://inch-ci.org/github/grempe/session-keys-rb)

`sessionKeys` is a cryptographic tool for the generation of unique user IDs,
and NaCl compatible [Curve25519](https://cr.yp.to/ecdh.html) encryption, and
[Ed25519](http://ed25519.cr.yp.to) digital signature keys using Ruby.

It is compatible with [grempe/session-keys-js](https://github.com/grempe/session-keys-js)
which can generates identical IDs and crypto keys using Javascript when given the
same username and passphrase values. Both libraries have extensive tests to
ensure they remain interoperable.

The strength of the system lies in the fact that the keypairs are derived from passing an identifier such as a username or email address, and a high-entropy passphrase through the SHA256 cryptographic one-way hash function, and then 'stretching' that username/password into strong key material using the scrypt key derivation function.

For an overview of the security design, please see the README for the companion
project [grempe/session-keys-js](https://github.com/grempe/session-keys-js)

## Installation

Add this line to your application's Gemfile:

```ruby
gem 'session_keys'
```

And then execute:

```text
$ bundle
```

Or install it yourself as:

```text
$ gem install session_keys
```

## Usage

```ruby
require 'session_keys'

SessionKeys.generate('user@example.com', 'my strong passphrase')

{
  id: '...',
  byte_keys: [...],
  hex_keys: [...],
  nacl_encryption_key_pairs: [...],
  nacl_encryption_key_pairs_base64: [...],
  nacl_signing_key_pairs: [...],
  nacl_signing_key_pairs_base64: [...],
  process_time: 250
}

```

Security Note : Each Array will contain eight values. Since each value at a
particular Array index is derived from the same key material it is recommended
to choose the different key types you need from different Array indexes. This
ensures that each key type was not derived from the same value.

```
# uuid : array index 0
output.hex_keys[0]

# encryption keypair : array index 1
output.nacl_encryption_key_pairs[1]

# signing keypair : array index 2
output.nacl_signing_key_pairs[2]
```

### Installation Security : Signed Git Commits

Most, if not all, of the commits and tags to the repository for this code are
signed with my PGP/GPG code signing key. I have uploaded my code signing public
keys to GitHub and you can now verify those signatures with the GitHub UI.
See [this list of commits](https://github.com/grempe/session-keys-rb/commits/master)
and look for the `Verified` tag next to each commit. You can click on that tag
for additional information.

You can also clone the repository and verify the signatures locally using your
own GnuPG installation. You can find my certificates and read about how to conduct
this verification at [https://www.rempe.us/keys/](https://www.rempe.us/keys/).

## Development

After checking out the repo, run `bin/setup` to install dependencies. Then,
run `rake test` to run the tests. You can also run `bin/console` for an
interactive prompt that will allow you to experiment.

To install this gem onto your local machine, run `bundle exec rake install`.

## Contributing

Bug reports and pull requests are welcome on GitHub at
[https://github.com/grempe/session-keys-rb](https://github.com/grempe/session-keys-rb).
This project is intended to be a safe, welcoming space for collaboration, and
contributors are expected to adhere to the
[Contributor Covenant](http://contributor-covenant.org) code of conduct.

## Legal

### Copyright

(c) 2016 Glenn Rempe <[glenn@rempe.us](mailto:glenn@rempe.us)> ([https://www.rempe.us/](https://www.rempe.us/))

### License

The gem is available as open source under the terms of
the [MIT License](http://opensource.org/licenses/MIT).

### Warranty

Unless required by applicable law or agreed to in writing,
software distributed under the License is distributed on an
"AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
either express or implied. See the LICENSE.txt file for the
specific language governing permissions and limitations under
the License.


CHANGELOG.md
# CHANGELOG

## v2.0.0 (8/23/2019)

- Update rbnacl. Now requires separate install of libsodium.
- Minor testing tweaks to modernize supported Ruby versions.
- Minor code climate lint cleanups.

## v1.0.0 (9/8/2016)

- Major refactor and simplification
- Compatible with grempe/session-keys-js

## v0.1.0 (4/30/2016)

This is the initial ALPHA quality release.

It is for review only and should not yet be used in production.


test/test_helper.rb
# coveralls.io
if ENV['TRAVIS'] || ENV['CI'] || ENV['JENKINS_URL'] || ENV['TDDIUM'] || ENV['COVERALLS_RUN_LOCALLY']
  # coveralls.io : web based code coverage
  require 'coveralls'
  Coveralls.wear!
end

$LOAD_PATH.unshift File.expand_path('../../lib', __FILE__)
require 'session_keys'

require 'minitest/autorun'
require 'minitest/pride'
require 'securerandom'


test/session_keys_test.rb
require_relative 'test_helper'

describe SessionKeys do
  describe 'constants' do
    it 'should be set to pre-determined interoperable values' do
      SessionKeys::SCRYPT_DIGEST_SIZE.must_equal 256
    end
  end

  describe 'id argument' do
    it 'must accept a min length 1 string' do
      keys = SessionKeys.generate('a', 'my secret password', 1)
      keys[:id].length.must_equal 64
    end

    it 'must accept a max length string' do
      keys = SessionKeys.generate('a'*256, 'my secret password', 1)
      keys[:id].length.must_equal 64
    end

    it 'must raise an ArgumentError if an invalid ID type was provided' do
      err = ->{ SessionKeys.generate(nil, 'my', 1) }.must_raise ArgumentError
      err.message.must_match(/invalid id, not a US-ASCII or UTF-8 string/)
      err = ->{ SessionKeys.generate('a'.force_encoding('BINARY'), 'my', 1) }.must_raise ArgumentError
      err.message.must_match(/invalid id, not a US-ASCII or UTF-8 string/)
    end

    it 'must raise an ArgumentError if an invalid ID length was provided' do
      err = ->{ SessionKeys.generate('', 'my', 1) }.must_raise ArgumentError
      err.message.must_match(/invalid id, must be between 1 and 256 characters in length/)
      err = ->{ SessionKeys.generate('a'*257, 'my', 1) }.must_raise ArgumentError
      err.message.must_match(/invalid id, must be between 1 and 256 characters in length/)
    end
  end

  describe 'password argument' do
    it 'must accept a min length 1 string as long as entropy is sufficient' do
      Zxcvbn.stub :test, OpenStruct.new({entropy: 1}) do
        keys = SessionKeys.generate('a', 'a', 1)
        keys.must_be_instance_of Hash
      end
    end

    it 'must accept a max length 256 string' do
      keys = SessionKeys.generate('a', SecureRandom.hex(128).force_encoding('UTF-8'), 512)
      keys.must_be_instance_of Hash
    end

    it 'must raise an ArgumentError if an invalid password type was provided' do
      err = ->{ SessionKeys.generate('someId', nil, 1) }.must_raise ArgumentError
      err.message.must_match(/invalid password, not a US-ASCII or UTF-8 string/)
      err = ->{ SessionKeys.generate('someId', 'bin'.force_encoding('BINARY'), 1) }.must_raise ArgumentError
      err.message.must_match(/invalid password, not a US-ASCII or UTF-8 string/)
    end

    it 'must raise an ArgumentError if an invalid password length was provided' do
      err = ->{ SessionKeys.generate('someId', '', 1) }.must_raise ArgumentError
      err.message.must_match(/invalid password, must be between 1 and 256 characters in length/)
      err = ->{ SessionKeys.generate('someId', 'a'*257, 1) }.must_raise ArgumentError
      err.message.must_match(/invalid password, must be between 1 and 256 characters in length/)
    end

    it 'must raise an ArgumentError if measured password entropy is below default minimum' do
      Zxcvbn.stub :test, OpenStruct.new({entropy: 74}) do
        err = ->{ SessionKeys.generate('someId', 'a') }.must_raise ArgumentError
        err.message.must_match(/invalid password, must be at least 75 bits of estimated entropy/)
      end
    end

    it 'must raise an ArgumentError if measured password entropy is below set minimum' do
      Zxcvbn.stub :test, OpenStruct.new({entropy: 9}) do
        err = ->{ SessionKeys.generate('someId', 'a', 10) }.must_raise ArgumentError
        err.message.must_match(/invalid password, must be at least 10 bits of estimated entropy/)
      end
    end
  end

  describe 'min_password_entropy argument' do
    it 'must accept minimum value of 1' do
      Zxcvbn.stub :test, OpenStruct.new({entropy: 1}) do
        keys = SessionKeys.generate('user@example.com', 'my secret password', 1)
        keys.must_be_instance_of Hash
      end
    end

    it 'must accept password with default value of 75' do
      Zxcvbn.stub :test, OpenStruct.new({entropy: 75}) do
        keys = SessionKeys.generate('user@example.com', 'my secret password')
        keys.must_be_instance_of Hash
      end
    end

    it 'must round the measured entropy to the nearest Integer' do
      Zxcvbn.stub :test, OpenStruct.new({entropy: 75.1}) do
        keys = SessionKeys.generate('user@example.com', 'my secret password', 75)
        keys.must_be_instance_of Hash
      end
    end

    it 'must accept maximum value of 512' do
      Zxcvbn.stub :test, OpenStruct.new({entropy: 512}) do
        keys = SessionKeys.generate('user@example.com', 'my secret password', 512)
        keys.must_be_instance_of Hash
      end
    end

    it 'must raise an ArgumentError if out of bounds' do
      err = ->{ SessionKeys.generate('someId', 'somePass', nil) }.must_raise ArgumentError
      err.message.must_match(/invalid min_password_entropy/)
      err = ->{ SessionKeys.generate('someId', 'somePass', -1) }.must_raise ArgumentError
      err.message.must_match(/invalid min_password_entropy/)
      err = ->{ SessionKeys.generate('someId', 'somePass', 0) }.must_raise ArgumentError
      err.message.must_match(/invalid min_password_entropy/)
      err = ->{ SessionKeys.generate('someId', 'somePass', '2') }.must_raise ArgumentError
      err.message.must_match(/invalid min_password_entropy/)
      err = ->{ SessionKeys.generate('someId', 'somePass', 513) }.must_raise ArgumentError
      err.message.must_match(/invalid min_password_entropy/)
    end
  end

  describe 'generate' do
    it 'must return the correct response hash' do
      keys = SessionKeys.generate('user@example.com', 'my secret password', 2)
      keys.keys.must_equal [:id, :byte_keys, :hex_keys, :nacl_encryption_key_pairs,
        :nacl_encryption_key_pairs_base64, :nacl_signing_key_pairs,
        :nacl_signing_key_pairs_base64, :process_time]
    end

    it 'must return id' do
      keys = SessionKeys.generate('user@example.com', 'my secret password', 2)
      keys.key?(:id).must_equal true
      keys[:id].must_be_instance_of String
      keys[:id].length.must_equal 64
    end

    it 'must return process_time of the right type and range' do
      keys = SessionKeys.generate('user@example.com', 'my secret password', 2)
      keys[:process_time].must_be_instance_of Float
      keys[:process_time].must_be :>, 10.0
      keys[:process_time].must_be :<, 100.0
    end
  end

  describe 'encrypt and decrypt' do
    it 'must work using generated encryption keys' do
      keys = SessionKeys.generate('user@example.com', 'my secret password', 2)

      # Assign the first of 8 keypairs to Alice for this test
      alice = keys[:nacl_encryption_key_pairs][0]

      # Bob gets the second set for this test
      bob = keys[:nacl_encryption_key_pairs][1]

      # The message Alice wants to send
      plaintext = 'foo bar baz'

      # Encrypt message from Alice to Bob
      a_box = RbNaCl::Box.new(bob[:public_key], alice[:secret_key])
      nonce = RbNaCl::Random.random_bytes(a_box.nonce_bytes)
      ciphertext = a_box.encrypt(nonce, plaintext)

      # Decryption of Alice's message by Bob
      # The nonce would be shared along with the ciphertext
      b_box = RbNaCl::Box.new(alice[:public_key], bob[:secret_key])
      decrypted_plaintext = b_box.open(nonce, ciphertext)

      decrypted_plaintext.must_equal plaintext
    end

    it 'can decrypt a message with RbNaCl encrypted by TweetNaCl.js client' do
      keys = SessionKeys.generate('user@example.com', 'my secret password', 2)

      alice = keys[:nacl_encryption_key_pairs][0]

      # Enter as the 'Their Public Key' value on website
      # alice[:public_key] = ONIMmjAJRAqO9iix85QojI03LCSC1O7miYSJA/23mlQ=
      alice_b64 = keys[:nacl_encryption_key_pairs_base64][0]

      # Created with : https://tweetnacl.js.org/#/sign
      nonce_b64 = 'KQIsPlYgOKGurOf1DuqlUTr4K8zr2M86'
      bob_secret_key_b64 = 'cds1gX0u9kTC76utKilzAk3kuKUTGwdUP0QmbHw4kCE='
      bob_public_key_b64 = 'LpwXfjbYZsReIKozHivmBPu8kPUEe7sXRjVQjhSD9nI='
      message = 'tweet-nacl-js'
      box_b64 = 'O/N+xSI/o9QSoS1V/9MPB78epwEvEiP79DdeUdQ='

      box = RbNaCl::Box.new(Base64.decode64(bob_public_key_b64), alice[:secret_key])
      decrypted_message = box.open(Base64.decode64(nonce_b64), Base64.decode64(box_b64))

      decrypted_message.must_equal message
    end
  end

  describe 'sign and verify' do
    it 'must work using generated signing keys' do
      keys = SessionKeys.generate('user@example.com', 'my secret password', 2)

      # Assign the first of 8 keypairs to Alice for this test
      alice = keys[:nacl_signing_key_pairs][0]

      # The message Alice wants to sign
      message = 'foo bar baz'

      signature = alice[:secret_key].sign(message)

      # Obtain the verify key for a given signing key
      verify_key = alice[:public_key]

      # Convert the verify key to a string to send it to a third party
      verify_key_s = verify_key.to_s

      # verify that the Base64 version this lib generates matches the Base64 encoded string output from RbNaCl
      # Base64.strict_encode64 is used which does not add a newline
      keys[:nacl_signing_key_pairs_base64][0][:public_key].must_equal Base64.strict_encode64(verify_key_s)

      # SEND MESSAGE + SIGNATURE TO BOB

      # Create a new verification key from the string
      verify_key = RbNaCl::VerifyKey.new(verify_key_s)

      # Check the validity of a message's signature
      # Will raise RbNaCl::BadSignatureError if the signature check fails
      verify_key.verify(signature, message).must_equal true
    end

    it 'can verify a message with RbNaCl signed by TweetNaCl.js client' do
      # Created with : https://tweetnacl.js.org/#/sign
      # secret key : nmZUa3RhbPJA8rJHj8B1UWtQp0ihM4mTE4M5XxGy/PE4ckiAHFabShY+Nfl01NGr5SxHplO5DRJo5hni/ynSBw==
      # public key : OHJIgBxWm0oWPjX5dNTRq+UsR6ZTuQ0SaOYZ4v8p0gc=
      # message : 'tweet-nacl-js'

      # generated signature
      verify_key_s = 'OHJIgBxWm0oWPjX5dNTRq+UsR6ZTuQ0SaOYZ4v8p0gc='

      # recipient verification
      verify_key = RbNaCl::VerifyKey.new(Base64.decode64(verify_key_s))
      verify_key.verify(Base64.decode64('SClbozQrNzjxS4rLZbFNYB1cu/EMosoOqZVlymLJ4o+ptjbeOSW+JS+vheJ5fQbcQm21jqRozfWBKkp7kePeBw=='), 'tweet-nacl-js').must_equal true
    end
  end

end


test/interoperability_test.rb
require_relative 'test_helper'

describe SessionKeys do

  ##########################################################################
  # WARNING : WARNING : WARNING : WARNING : WARNING : WARNING : WARNING
  # DO NOT CHANGE THE RESULT VALUES. THEY ARE USED BY MORE THAN ONE LIBRARY
  # AND THE OUTPUTS MUST BE CONSTANT TO ENSURE INTEROPERABILITY!
  #
  # IF THESE TESTS NO LONGER PASS THEN INTEROPERABILITY IS BROKEN!
  ##########################################################################

  describe 'interop test' do
    it 'must create the expected output' do
      keys = SessionKeys.generate('user@example.com', 'pet sprain our trial patch bg')

      # ID : SHA256 of ID
      keys[:id].must_equal 'b4c9a289323b21a01c3e940f150eb9b8c542587f1abfd8f0e1cc1ffc5e475514'

      keys[:byte_keys].must_equal [
        [227, 213, 226, 155, 22, 212, 112, 32, 60, 247, 86, 101, 125, 154, 151, 234, 71, 59, 82, 79, 8, 52, 188, 100, 91, 28, 165, 216, 183, 137, 157, 17],
        [1, 37, 254, 86, 67, 74, 79, 174, 191, 132, 239, 255, 168, 102, 235, 106, 29, 219, 45, 47, 237, 108, 62, 205, 151, 27, 163, 160, 14, 0, 166, 54],
        [81, 64, 179, 222, 199, 221, 82, 29, 93, 162, 162, 48, 44, 38, 209, 199, 21, 244, 64, 92, 47, 201, 177, 111, 93, 89, 82, 130, 203, 120, 135, 187],
        [128, 57, 78, 134, 230, 191, 121, 88, 50, 177, 118, 75, 63, 231, 21, 168, 106, 103, 187, 78, 254, 58, 140, 198, 237, 3, 109, 126, 68, 60, 204, 216],
        [125, 57, 175, 89, 168, 203, 202, 97, 200, 211, 78, 174, 118, 117, 162, 77, 206, 120, 124, 239, 76, 158, 219, 104, 27, 72, 253, 129, 100, 216, 68, 122],
        [244, 123, 137, 212, 254, 81, 59, 36, 159, 247, 79, 163, 24, 189, 249, 58, 104, 13, 58, 174, 84, 236, 166, 53, 158, 251, 235, 160, 188, 44, 17, 35],
        [70, 41, 248, 98, 4, 156, 146, 253, 236, 23, 38, 177, 1, 91, 139, 123, 15, 96, 53, 41, 168, 60, 244, 52, 89, 16, 219, 60, 29, 183, 32, 110],
        [61, 52, 141, 115, 90, 229, 18, 231, 253, 192, 39, 20, 196, 222, 98, 126, 178, 56, 26, 30, 100, 75, 225, 191, 81, 74, 155, 41, 78, 19, 53, 97]
      ]

      keys[:hex_keys].must_equal [
        'e3d5e29b16d470203cf756657d9a97ea473b524f0834bc645b1ca5d8b7899d11',
        '0125fe56434a4faebf84efffa866eb6a1ddb2d2fed6c3ecd971ba3a00e00a636',
        '5140b3dec7dd521d5da2a2302c26d1c715f4405c2fc9b16f5d595282cb7887bb',
        '80394e86e6bf795832b1764b3fe715a86a67bb4efe3a8cc6ed036d7e443cccd8',
        '7d39af59a8cbca61c8d34eae7675a24dce787cef4c9edb681b48fd8164d8447a',
        'f47b89d4fe513b249ff74fa318bdf93a680d3aae54eca6359efbeba0bc2c1123',
        '4629f862049c92fdec1726b1015b8b7b0f603529a83cf4345910db3c1db7206e',
        '3d348d735ae512e7fdc02714c4de627eb2381a1e644be1bf514a9b294e133561'
      ]

      keys[:nacl_encryption_key_pairs].size.must_equal 8
      keys[:nacl_encryption_key_pairs].first.must_be_instance_of Hash
      keys[:nacl_encryption_key_pairs].first[:secret_key].must_be_instance_of RbNaCl::Boxes::Curve25519XSalsa20Poly1305::PrivateKey
      keys[:nacl_encryption_key_pairs].first[:public_key].must_be_instance_of RbNaCl::Boxes::Curve25519XSalsa20Poly1305::PublicKey

      keys[:nacl_encryption_key_pairs_base64].must_equal [
        {:secret_key=>'49XimxbUcCA891ZlfZqX6kc7Uk8INLxkWxyl2LeJnRE=', :public_key=>'9G8XJgiIXj32stQmxtUa8vmmvLGTssTrEwd9tIYpVkA='},
        {:secret_key=>'ASX+VkNKT66/hO//qGbrah3bLS/tbD7NlxujoA4ApjY=', :public_key=>'JH73SmhYv43j25rpC8q797XHQi4hx/DrAcQCb5i143k='},
        {:secret_key=>'UUCz3sfdUh1doqIwLCbRxxX0QFwvybFvXVlSgst4h7s=', :public_key=>'HmpoJqIjMnHYqvQheiCc8HXymyiGHX3ell8A+2WE330='},
        {:secret_key=>'gDlOhua/eVgysXZLP+cVqGpnu07+OozG7QNtfkQ8zNg=', :public_key=>'otEJhqs+cpZ1x4OHva30k4T8ye7x5eUBc2UR5IcZkhc='},
        {:secret_key=>'fTmvWajLymHI006udnWiTc54fO9MnttoG0j9gWTYRHo=', :public_key=>'AHHyBuknbD8/2sOe5fUWY7y9zVkOD1SrjaLdBRglVDM='},
        {:secret_key=>'9HuJ1P5ROySf90+jGL35OmgNOq5U7KY1nvvroLwsESM=', :public_key=>'uFKZeT1VgkltnczCmljitEQswPFjQT2mS6nRKAJ4YnE='},
        {:secret_key=>'Rin4YgSckv3sFyaxAVuLew9gNSmoPPQ0WRDbPB23IG4=', :public_key=>'W/VA/NmgV4idPxVtzuGE8Uo5bs2fQU0L2cxhoZJFgyU='},
        {:secret_key=>'PTSNc1rlEuf9wCcUxN5ifrI4Gh5kS+G/UUqbKU4TNWE=', :public_key=>'zyQcH+swlgDiBwzDAtFnTMKWR//DdNPyFYNLLMOLsjc='}
      ]

      keys[:nacl_signing_key_pairs].size.must_equal 8
      keys[:nacl_signing_key_pairs].first.must_be_instance_of Hash
      keys[:nacl_signing_key_pairs].first[:secret_key].must_be_instance_of RbNaCl::Signatures::Ed25519::SigningKey
      keys[:nacl_signing_key_pairs].first[:public_key].must_be_instance_of RbNaCl::Signatures::Ed25519::VerifyKey

      keys[:nacl_signing_key_pairs_base64].must_equal [
        {:secret_key=>'49XimxbUcCA891ZlfZqX6kc7Uk8INLxkWxyl2LeJnRE=', :public_key=>'9mc6TmR6fw+OfPZA+TI4pDMeensYo3vHjCAWwNJr5Sg='},
        {:secret_key=>'ASX+VkNKT66/hO//qGbrah3bLS/tbD7NlxujoA4ApjY=', :public_key=>'IA4yoeU/2xv2elvmJWFLP3Hiy2Hp5FdGpdrJjkf+5FU='},
        {:secret_key=>'UUCz3sfdUh1doqIwLCbRxxX0QFwvybFvXVlSgst4h7s=', :public_key=>'/orhEpXoWrbHQcWlg/IEnNiJvW+2j0lS7+/gvOFlppc='},
        {:secret_key=>'gDlOhua/eVgysXZLP+cVqGpnu07+OozG7QNtfkQ8zNg=', :public_key=>'JAr8Pcij0HTvraNF9UeZ3vx0rvtixv4aIIMuDXrq+xc='},
        {:secret_key=>'fTmvWajLymHI006udnWiTc54fO9MnttoG0j9gWTYRHo=', :public_key=>'j9v+uOiZ2iVYwIiSMEeh8LVtFVagKkQ0n8lM6g7NIEY='},
        {:secret_key=>'9HuJ1P5ROySf90+jGL35OmgNOq5U7KY1nvvroLwsESM=', :public_key=>'SJF1MvZ0Ggb1UQWHKkn9NHAkHU9A/ofV159fCsB6Pbo='},
        {:secret_key=>'Rin4YgSckv3sFyaxAVuLew9gNSmoPPQ0WRDbPB23IG4=', :public_key=>'KlYHO13rFU3vlWxHvMo0s7nILVH6rzsgDAblSz3yVaw='},
        {:secret_key=>'PTSNc1rlEuf9wCcUxN5ifrI4Gh5kS+G/UUqbKU4TNWE=', :public_key=>'HllzWBHwzXC4XnQU0xSzGDYN5aUhD2lbSQJVF3f2mQA='}
      ]
    end
  end
end


lib/session_keys.rb
# frozen_string_literal: true

require 'session_keys/version'
require 'rbnacl'
require 'zxcvbn'
require 'base64'

# SessionKeys deterministic cryptographic key generation.
module SessionKeys
  # Size in bytes of the scrypt derived output
  SCRYPT_DIGEST_SIZE = 256

  # Deterministically generates a collection of derived encryption key material from
  # a provided id and password/passphrase. Uses SHA256 and scrypt for key derivation.
  #
  # @param id [String] a unique US-ASCII or UTF-8 encoded String identifier such as a username or
  #   email address. Max length 256 characters.
  # @param password [String] a cryptographically strong US-ASCII or UTF-8 encoded password or
  #   passphrase. Max length 256 characters.
  # @param min_password_entropy [Integer] the minimum (75) estimated entropy allowed
  #   for the password. This will be measured with Zxcvbn.
  # @return [Hash] returns a Hash of keys and derived key material.
  # @raise [ArgumentError] if invalid arguments are provided.
  def self.generate(id, password, min_password_entropy = 75)
    unless id.is_a?(String) && ['US-ASCII', 'UTF-8'].include?(id.encoding.name)
      raise ArgumentError, 'invalid id, not a US-ASCII or UTF-8 string'
    end

    unless id.length.between?(1,256)
      raise ArgumentError, 'invalid id, must be between 1 and 256 characters in length'
    end

    unless password.is_a?(String) && ['US-ASCII', 'UTF-8'].include?(password.encoding.name)
      raise ArgumentError, 'invalid password, not a US-ASCII or UTF-8 string'
    end

    # Enforce max length only due to Zxcvbn taking a *long* time to
    # process long strings and determine entropy.
    unless password.length.between?(1,256)
      raise ArgumentError, 'invalid password, must be between 1 and 256 characters in length'
    end

    unless min_password_entropy.is_a?(Integer) && min_password_entropy.between?(1, 512)
      raise ArgumentError, 'invalid min_password_entropy, must be an Integer between 1 and 512'
    end

    password_test = Zxcvbn.test(password)
    unless password_test.entropy.round >= min_password_entropy
      raise ArgumentError, "invalid password, must be at least #{min_password_entropy} bits of estimated entropy"
    end

    start_processing_time = Time.now

    id_sha256_bytes = RbNaCl::Hash.sha256(id.bytes.pack('C*'))
    id_sha256_hex = id_sha256_bytes.bytes.map { |byte| '%02x' % byte }.join

    # libsodium : By design, a password whose length is 65 bytes or more is
    # reduced to SHA-256(password). This can have security implications if the
    # password is present in another password database using raw, unsalted
    # SHA-256. Or when upgrading passwords previously hashed with unsalted
    # SHA-256 to scrypt. If this is a concern, passwords should be pre-hashed
    # before being hashed using scrypt.
    scrypt_key = RbNaCl::Hash.sha256(password.bytes.pack('C*'))

    # Tie the sycrypt password bytes to the ID they are associate with by
    # utilizing the ID as the salt. Include the ID length and an additional
    # string to harden the salt.
    scrypt_salt = RbNaCl::Hash.sha256([id_sha256_hex,
                                       id_sha256_hex.length,
                                       'session_keys'].join('').bytes.pack('C*'))

    # Derive SCRYPT_DIGEST_SIZE secret bytes
    password_digest = RbNaCl::PasswordHash.scrypt(
      scrypt_key,
      scrypt_salt,
      16384 * 32,
      16384 * 32 * 32,
      SCRYPT_DIGEST_SIZE
    ).bytes

    num_keys = SCRYPT_DIGEST_SIZE / 32

    byte_keys = []
    num_keys.times { byte_keys << password_digest.shift(32) }

    hex_keys = byte_keys.map { |key|
      key.map { |byte| '%02x' % byte }.join
    }

    nacl_encryption_key_pairs = byte_keys.map { |key|
      seed = key.pack('C*').force_encoding('ASCII-8BIT')
      sec_key = RbNaCl::PrivateKey.new(seed)
      pub_key = sec_key.public_key
      {secret_key: sec_key, public_key: pub_key}
    }

    nacl_encryption_key_pairs_base64 = nacl_encryption_key_pairs.map { |keypair|
      pub_key = Base64.strict_encode64(keypair[:public_key].to_bytes)
      sec_key = Base64.strict_encode64(keypair[:secret_key].to_bytes)
      {secret_key: sec_key, public_key: pub_key}
    }

    nacl_signing_key_pairs = byte_keys.map { |key|
      seed = key.pack('C*').force_encoding('ASCII-8BIT')
      sec_key = RbNaCl::SigningKey.new(seed)
      pub_key = sec_key.verify_key
      {secret_key: sec_key, public_key: pub_key}
    }

    nacl_signing_key_pairs_base64 = nacl_signing_key_pairs.map { |keypair|
      pub_key = Base64.strict_encode64(keypair[:public_key].to_bytes)
      sec_key = Base64.strict_encode64(keypair[:secret_key].to_bytes)
      {secret_key: sec_key, public_key: pub_key}
    }

    {
      id: id_sha256_hex,
      byte_keys: byte_keys,
      hex_keys: hex_keys,
      nacl_encryption_key_pairs: nacl_encryption_key_pairs,
      nacl_encryption_key_pairs_base64: nacl_encryption_key_pairs_base64,
      nacl_signing_key_pairs: nacl_signing_key_pairs,
      nacl_signing_key_pairs_base64: nacl_signing_key_pairs_base64,
      process_time: ((Time.now - start_processing_time)*1000).round(2)
    }
  end
end
